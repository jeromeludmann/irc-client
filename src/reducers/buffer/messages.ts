import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RAW_MESSAGES_RECEIVED,
  RawMessagesReceivedAction,
  CONNECTION_CLOSED,
  ConnectionClosedAction,
} from "@app/actions/network";
import {
  IncomingJoinAction,
  IncomingServerPingAction,
  JOIN,
  SERVER_NOTICE,
  ERROR,
  IncomingErrorAction,
  IncomingPrivmsgAction,
  PRIVMSG,
  IncomingServerNoticeAction,
  SERVER_PING,
  PART,
  IncomingPartAction,
} from "@app/actions/incoming";

export type MessageListState = string[];

export type MessageListAction =
  | ConnectionClosedAction
  | ConnectionFailedAction
  | IncomingErrorAction
  | IncomingJoinAction
  | IncomingPartAction
  | IncomingPrivmsgAction
  | IncomingServerNoticeAction
  | IncomingServerPingAction
  | RawMessagesReceivedAction;

export const messagesInitialState: MessageListState = [];

export default (
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState => {
  switch (action.type) {
    case CONNECTION_FAILED: {
      return [...messages, action.payload.message];
    }

    case CONNECTION_CLOSED: {
      return [...messages, "Disconnected from remote host."];
    }

    case ERROR: {
      return [...messages, action.payload.message];
    }

    case JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }

    case PART: {
      const { user, channel, message } = action.payload;
      const baseMsg = `${user.nick} has left ${channel}`;
      const msg = message ? `${baseMsg} (${message})` : baseMsg;
      return [...messages, msg];
    }

    case PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }

    case RAW_MESSAGES_RECEIVED: {
      return [...messages, ...action.payload.messages];
    }

    case SERVER_NOTICE: {
      return [...messages, action.payload.text];
    }

    case SERVER_PING: {
      return [...messages, "Ping?"];
    }

    default:
      return messages;
  }
};
