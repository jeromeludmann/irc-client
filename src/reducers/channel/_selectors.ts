import { createSelector } from "reselect";
import { selectChannel } from "@app/reducers/server/_selectors";

export const selectMessages = createSelector(
  selectChannel,
  channel => channel.messages,
);

export const selectInput = createSelector(selectChannel, channel => channel.input);
