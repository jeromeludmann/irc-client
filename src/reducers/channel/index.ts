import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from "@app/reducers/channel/messages";
import {
  InputState,
  inputInitialState,
  reduceInput,
} from "@app/reducers/input";
import {
  UnreadState,
  unreadInitialState,
  reduceUnread,
} from "@app/reducers/channel/unread";
import { RoutedAction } from "@app/Route";
import { RouteState } from "@app/reducers/route";
import { UserState } from "@app/reducers/server/user";

export interface ChannelState {
  readonly messages: MessagesState;
  readonly input: InputState;
  readonly unread: UnreadState;
}

export const channelInitialState: ChannelState = {
  messages: messagesInitialState,
  input: inputInitialState,
  unread: unreadInitialState,
};

export const reduceChannel = (
  channel = channelInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
): ChannelState => ({
  messages: reduceMessages(channel.messages, action),
  input: reduceInput(channel.input, action),
  unread: reduceUnread(channel.unread, action, extraStates),
});
