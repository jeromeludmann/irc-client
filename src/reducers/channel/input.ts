import { Reducer, Action } from "redux";
import {
  UPDATE_INPUT_VALUE,
  UpdateInputValueAction,
  ENTER_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
  EnterInputValueAction,
} from "@app/actions/ui";

export interface InputState {
  value: string;
  dirtyValue: string;
  history: {
    values: string[];
    index: number;
  };
}

export const inputInitialState: InputState = {
  value: "",
  dirtyValue: "",
  history: { values: [], index: 0 },
};

const beginOfHistory = ({ history }: InputState) => history.index === 0;

const endOfHistory = ({ history }: InputState) =>
  history.index === history.values.length;

const handlers: {
  [action: string]: (input: InputState, action: Action) => InputState;
} = {
  [UPDATE_INPUT_VALUE]: (input, action: UpdateInputValueAction) => ({
    ...input,
    value: action.payload.value,
    dirtyValue: endOfHistory(input) ? action.payload.value : input.dirtyValue,
  }),

  [ENTER_INPUT_VALUE]: (input, action: EnterInputValueAction) => {
    const historyValues = [...input.history.values, action.payload.value];
    return {
      value: "",
      dirtyValue: "",
      history: { values: historyValues, index: historyValues.length },
    };
  },

  [GO_BACK_INPUT_HISTORY]: input =>
    beginOfHistory(input)
      ? input
      : {
          ...input,
          value: input.history.values[input.history.index - 1],
          history: { ...input.history, index: input.history.index - 1 },
        },

  [GO_FORWARD_INPUT_HISTORY]: input =>
    endOfHistory(input)
      ? input
      : {
          ...input,
          value:
            input.history.values[input.history.index + 1] || input.dirtyValue,
          history: { ...input.history, index: input.history.index + 1 },
        },
};

export const reduceInput: Reducer<InputState, Action> = (
  input = inputInitialState,
  action,
) =>
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](input, action)
    : input;
