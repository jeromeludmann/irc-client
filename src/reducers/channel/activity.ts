import { RoutedAction, BROADCAST_ACTIVE } from "@app/Route";
import { RouteState } from "@app/reducers/route";
import { SWITCH_WINDOW } from "@app/actions/ui";
import { UserState } from "@app/reducers/server/user";
import { JoinAction, JOIN } from "@app/actions/messages";

export type ActivityState = boolean;

export const activityInitialState: ActivityState = false;

type ActivityReducer<A = RoutedAction> = (
  state: ActivityState,
  action: A,
  extraStates: { route: RouteState; user: UserState },
) => ActivityState;

const reducers: { [action: string]: ActivityReducer } = {
  [JOIN]: (_, action: JoinAction, extraStates) =>
    action.payload.user.nick !== extraStates.user.nick,

  [SWITCH_WINDOW]: () => false,
};

const hasAnyActivity: ActivityReducer<RoutedAction> = (
  _,
  action,
  extraStates,
) => {
  const sameRoute =
    action.route.serverKey === extraStates.route.serverKey &&
    action.route.channelKey === extraStates.route.channelKey;
  return !sameRoute && action.route.channelKey !== BROADCAST_ACTIVE;
};

export const reduceActivity = (
  activity: ActivityState = activityInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
) =>
  reducers.hasOwnProperty(action.type)
    ? reducers[action.type](activity, action, extraStates)
    : hasAnyActivity(activity, action, extraStates);
