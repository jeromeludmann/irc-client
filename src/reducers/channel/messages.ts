import { Action } from "redux";
import { mapReducer } from "@app/reducers/_map";
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  ConnectionClosedAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
} from "@app/actions/socket";
import {
  ErrorAction,
  JoinAction,
  PartAction,
  PrivmsgAction,
  NoticeFromServerAction,
  NoticeFromChannelAction,
  NoticeFromUserAction,
  PingFromServerAction,
  ERROR,
  JOIN,
  NOTICE_FROM_SERVER,
  NOTICE_FROM_CHANNEL,
  NOTICE_FROM_USER,
  PART,
  PRIVMSG,
  PING_FROM_SERVER,
} from "@app/actions/messages";

export type MessagesState = string[];

export const messagesInitialState: MessagesState = [];

type MessagesReducer<A = Action> = (
  messages: MessagesState,
  action: A,
) => MessagesState;

const connectionFailed: MessagesReducer<ConnectionFailedAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.message];
};

const connectionClosed: MessagesReducer<ConnectionClosedAction> = messages => {
  return [...messages, "Disconnected from remote host."];
};

const error: MessagesReducer<ErrorAction> = (messages, action) => {
  return [...messages, action.payload.message];
};

const join: MessagesReducer<JoinAction> = (messages, action) => {
  const { user, channel } = action.payload;
  const msg = `${user.nick} has joined ${channel}`;
  return [...messages, msg];
};

const part: MessagesReducer<PartAction> = (messages, action) => {
  const { user, channel, message } = action.payload;
  const baseMsg = `${user.nick} has left ${channel}`;
  const msg = message ? `${baseMsg} (${message})` : baseMsg;
  return [...messages, msg];
};

const privmsg: MessagesReducer<PrivmsgAction> = (messages, action) => {
  const { user, text } = action.payload;
  const msg = `${user.nick}: ${text}`;
  return [...messages, msg];
};

const raw: MessagesReducer<RawMessagesAction> = (messages, action) => {
  return [...messages, ...action.payload.messages];
};

const noticeFromServer: MessagesReducer<NoticeFromServerAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.text];
};

const noticeFromChannel: MessagesReducer<NoticeFromChannelAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  const { channelKey } = action.route;
  return [...messages, `-${user.nick}/${channelKey}- ${text}`];
};

const noticeFromUser: MessagesReducer<NoticeFromUserAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  return [...messages, `-${user.nick}- ${text}`];
};

const serverPing: MessagesReducer<PingFromServerAction> = messages => {
  return [...messages, "Ping?"];
};

const map: { [action: string]: MessagesReducer } = {
  [CONNECTION_FAILED]: connectionFailed,
  [CONNECTION_CLOSED]: connectionClosed,
  [ERROR]: error,
  [JOIN]: join,
  [NOTICE_FROM_SERVER]: noticeFromServer,
  [NOTICE_FROM_CHANNEL]: noticeFromChannel,
  [NOTICE_FROM_USER]: noticeFromUser,
  [PART]: part,
  [PRIVMSG]: privmsg,
  [RAW_MESSAGES_RECEIVED]: raw,
  [PING_FROM_SERVER]: serverPing,
};

export const reduceMessages = mapReducer<MessagesState>(map);
