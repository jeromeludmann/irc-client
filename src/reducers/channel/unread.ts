import { RouteState } from "@app/reducers/route";
import { SWITCH_WINDOW, SwitchWindowAction } from "@app/actions/ui";
import { RAW_MESSAGES_RECEIVED } from "@app/actions/socket";
import { UserState } from "@app/reducers/server/user";
import { mapReducer } from "@app/reducers/_map";
import { Action } from "redux";
import { RoutedAction, BROADCAST_ACTIVE } from "@app/Route";
import {
  JoinAction,
  JOIN,
  PART,
  PRIVMSG,
  NOTICE_FROM_SERVER,
  NOTICE_FROM_CHANNEL,
  NOTICE_FROM_USER,
} from "@app/actions/messages";

export type UnreadState = boolean;

export const unreadInitialState: UnreadState = false;

type UnreadReducer<A = Action> = (
  state: UnreadState,
  action: A,
  extraStates: { route: RouteState; user: UserState },
) => UnreadState;

const join: UnreadReducer<JoinAction> = (_, action, extraStates) =>
  action.payload.user.nick !== extraStates.user.nick;

const switchRoute: UnreadReducer<SwitchWindowAction> = () => false;

const anyActivity: UnreadReducer<RoutedAction> = (_, action, extraStates) => {
  const sameRoute =
    action.route.serverKey === extraStates.route.serverKey &&
    action.route.channelKey === extraStates.route.channelKey;
  return !sameRoute && action.route.channelKey !== BROADCAST_ACTIVE;
};

const map: { [action: string]: UnreadReducer } = {
  [JOIN]: join,
  [SWITCH_WINDOW]: switchRoute,
  [PART]: anyActivity,
  [PRIVMSG]: anyActivity,
  [NOTICE_FROM_SERVER]: anyActivity,
  [NOTICE_FROM_CHANNEL]: anyActivity,
  [NOTICE_FROM_USER]: anyActivity,
  [RAW_MESSAGES_RECEIVED]: anyActivity,
};

export const reduceUnread = mapReducer<
  UnreadState,
  { route: RouteState; user: UserState }
>(map);
