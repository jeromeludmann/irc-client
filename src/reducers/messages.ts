import { Action } from 'redux'
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
  CONNECT_TO_SERVER,
  ConnectToServerAction,
} from '@app/actions/socket'
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
  RECEIVE_PONG_FROM_SERVER,
  ReceivePongFromServerAction,
} from '@app/actions/msgIncoming'
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from '@app/actions/msgOutgoing'
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from '@app/actions/commands'
import { RouteState } from '@app/reducers/route'
import { ServerState } from '@app/reducers/server'
import { CaseReducerMap } from '@app/utils/CaseReducerMap'

// TODO replace string[] by MessageState[]
export type MessagesState = Readonly<string[]>

type MessagesReducer = (
  messages: MessagesState,
  action: Action,
  extraStates: { route: RouteState; server: ServerState },
) => MessagesState

export const messagesInitialState: MessagesState = []

const caseReducers: CaseReducerMap<MessagesReducer> = {
  [CONNECT_TO_SERVER]: (messages, action: ConnectToServerAction) => [
    ...messages,
    `Connecting to ${action.payload.host}â€¦`,
  ],

  [CONNECTION_CLOSED]: messages => [
    ...messages,
    'Disconnected from remote host.',
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    'Type /HELP <command> for more details.',
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { bufferKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${bufferKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg]
  },

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, 'Ping?'],

  [RECEIVE_PONG_FROM_SERVER]: (
    messages,
    { payload: { key, lag } }: ReceivePongFromServerAction,
    {},
  ) => [...messages, `Response from server ${key} in ${lag} ms`],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [SEND_PONG_TO_SERVER]: messages => [...messages, 'Pong!'],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { server }) => [
    ...messages,
    `${server.user.nick}: ${action.payload.text}`,
  ],
}

export const reduceMessages: MessagesReducer = (
  messagesState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](messagesState, action, extraStates)
    : messagesState
