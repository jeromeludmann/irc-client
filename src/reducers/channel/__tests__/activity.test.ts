import { reduceActivity } from "@app/reducers/channel/activity";
import { messagesReceived } from "@app/actions/messages";
import { switchWindow } from "@app/actions/ui";

describe("reduce channel activity", () => {
  const extraParams = {
    user: { nick: "nick", user: "user", real: "name" },
    route: { serverKey: "server1", channelKey: "#channel" },
  };

  test("by joining channel", () => {
    expect(
      reduceActivity(
        undefined,
        messagesReceived["JOIN"](
          "server1",
          {
            nick: "nick",
            user: "user",
            host: "host",
          },
          ["#channel"],
        ),
        extraParams,
      ),
    ).toMatchSnapshot();
  });

  test("by switching window", () => {
    expect(
      reduceActivity(
        undefined,
        switchWindow({ serverKey: "server1", channelKey: "#channel" }),
        extraParams,
      ),
    ).toMatchSnapshot();
  });
});
