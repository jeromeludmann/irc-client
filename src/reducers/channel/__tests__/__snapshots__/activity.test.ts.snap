// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce channel activity by joining channel 1`] = `false`;

exports[`reduce channel activity by switching window 1`] = `false`;
